### PR TITLE
Add JS interface for pact

### PR DESCRIPTION
**Pull request by surangap**
## Changelog
- Add JS interface lib for pact. library will be generated using `wasm-pack`

### Crate
- https://crates.io/crates/trn-pact

### NPM package
- https://www.npmjs.com/package/@therootnetwork/pact-js

### CODE REVIEW

File: js/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+[package]
+ name = "js"
+ version = "0.0.1"
+ authors = ["The Root Network Team"]
+ edition = "2021"
+ description = "JS/Wasm bindings for the rust TRNNut codec"
```
Comment (JasonTulp) [inline js/Cargo.toml:6]: nit
```suggestion
 description = "JS/Wasm bindings for the rust TRN-Pact codec"
```

File: js/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Pact
+
+TODO
```
Comment (JasonTulp) [inline js/README.md:3]: TODO

**Review by JasonTulp — APPROVED** [review]

File: fuzz/fuzz_targets/contract_v0_fuzz.rs
Comment (zees-dev) [inline fuzz/fuzz_targets/contract_v0_fuzz.rs:1]: Do we need this file; only contains a function with a single method; maybe we can just call that function directly where needed?

File: js/test/pact.spec.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        ]);
+
+        let pactContract = new Pact(data_table, bytecode);
+        expect(pactContract.getBytecode()).toMatchObject(bytecode);
+        let encoded_pact = pactContract.encode();
+    });
+
+    // refer to ../../src/types/contract.rs "contract_encode_1" test for rust side output
+    test("pact contract encode same as rust output 1", () => {
+        let expected_payload = new Uint8Array([
```
Comment (zees-dev) [inline js/test/pact.spec.js:24]: we need to know where these values are coming from; how this is derived.
If the test ever breaks, we know exactly why.
If possible, the unit8array should be derived in the test itself.

File: src/types/base.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ use bit_reverse::ParallelReverse;
 /// A string-like type
 #[cfg_attr(feature = "std", derive(Debug))]
```
Comment (zees-dev) [inline src/types/base.rs:24]: debug doesn't really need `std`

File: src/types/base.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ use bit_reverse::ParallelReverse;
...
 #[derive(PartialEq, PartialOrd, Clone)]
-pub struct StringLike<'a>(pub &'a [u8]);
+pub struct StringLike(pub Vec<u8>);
 
 /// A numeric type
 #[cfg_attr(feature = "std", derive(Debug))]
```
Comment (zees-dev) [inline src/types/base.rs:29]: same for all the below; we don't need to feature flag just for debug trait

File: src/types/contract.rs [hunk 6]
```diff
@@ -68,17 +68,92 @@ impl<'a> Contract<'a> {
...
+            ]),
+            bytecode: vec![
+                OpCode::COMP(Comparator::new(OpComp::EQ)).into(),
+                0x00,
+                OpCode::CONJ(Conjunction::new(OpConj::AND)).into(),
+            ],
+        };
+        let mut encoded_payload = vec![];
+        contract.encode(&mut encoded_payload);
+        println!("{:?}", encoded_payload);
```
Comment (zees-dev) [inline src/types/contract.rs:157]: we should be asserting whether the encoded output matches some expectation

File: src/types/data_table.rs [hunk 7]
```diff
@@ -20,15 +20,15 @@ use bit_reverse::ParallelReverse;
 
 /// A pact contract's static data table
 #[cfg_attr(feature = "std", derive(PartialEq, Debug))]
```
Comment (zees-dev) [inline src/types/data_table.rs:22]: same here; don't need i guess

File: src/lib.rs [hunk 8]
```diff
@@ -11,28 +11,16 @@
 
 // You should have received a copy of the GNU General Public License
 // along with Pact. If not, see:
-//   <https://centrality.ai/licenses/gplv3.txt>
-//   <https://centrality.ai/licenses/lgplv3.txt>
+//   <https://futureverse.com/licenses/gplv3.txt>
+//   <https://futureverse.com/licenses/lgplv3.txt>
```
Comment (zees-dev) [inline src/lib.rs:15]: using apache now

File: README.md
Comment (zees-dev) [inline README.md:1]: probably need to be updated too :/

**Review by zees-dev — APPROVED** [review]
Few quality of life improvements required; however not a blocker.

LGTM 👍

File: js/test/pact.spec.js (hunk 3 above)
Comment (surangap) [reply js/test/pact.spec.js:24]: these tests check the js side encode output against the output of the same rust side test. references have been updated for the remaining tests for where to find the rust side test.

File: src/types/base.rs (hunk 4 above)
Comment (surangap) [reply src/types/base.rs:24]: it's more like when `std` is defined then only derive the `Debug`. Tried to remove this, leads to errors. might require removing it from all the references. Keeping it as it is for now.

File: src/types/contract.rs (hunk 6 above)
Comment (surangap) [reply src/types/contract.rs:157]: These generated outputs will be used in the js side tests as referred here ->

File: js/README.md (hunk 2 above)
Comment (surangap) [reply js/README.md]: [c274f9b](https://github.com/futureversecom/trn-pact/pull/3/commits/c274f9b8a66f7eb09eef124756f7bb2eed302f4e)

File: js/test/pact.spec.js (hunk 3 above)
Comment (surangap) [reply js/test/pact.spec.js:24]: [8c42e0e](https://github.com/futureversecom/trn-pact/pull/3/commits/8c42e0ec498af10b7584c67cc1324d4720b2745f)

File: src/lib.rs (hunk 8 above)
Comment (surangap) [reply src/lib.rs]: [888bafc](https://github.com/futureversecom/trn-pact/pull/3/commits/888bafc61bd3774ebb16bc74a2787e37a2d5a13b)

File: README.md
Comment (surangap) [reply README.md:1]: [23a1fd3](https://github.com/futureversecom/trn-pact/pull/3/commits/23a1fd34b7ee2ccdaf928fbab492797834d7be21)

**Pull request merged**